odoo/addons/base/models/ir_cron.py

```diff
@@ -291,82 +306,147 @@ def _acquire_one_job(cls, cr, job_ids):
-        #  --|-----|-----|-----|-----|-----|-----|----> time
-        #    1     2*    *     *     *  3  4
-        #
-        # 1: lastcall, the last time the cron was executed
-        # 2: past_nextcall, the cron nextcall as seen from lastcall
-        # *: missed_call, a total of 4 calls are missing
-        # 3: now
-        # 4: future_nextcall, the cron nextcall as seen from now
+        cron_cr.execute("""
+            UPDATE ir_cron
+            SET failures = %s,
+                first_failure_date = %s,
+                active = %s
+            WHERE id = %s
+        """, [
+            job['failures'],
+            job['first_failure_date'],
+            job['active'],
+            job['id'],
+        ])
 
-        with cls.pool.cursor() as job_cr:
-            lastcall = fields.Datetime.to_datetime(job['lastcall'])
-            interval = _intervalTypes[job['interval_type']](job['interval_number'])
-            env = api.Environment(job_cr, job['user_id'], {'lastcall': lastcall})
-            ir_cron = env[cls._name]
-
-            # Use the user's timezone to compare and compute datetimes,
-            # otherwise unexpected results may appear. For instance, adding
-            # 1 month in UTC to July 1st at midnight in GMT+2 gives July 30
-            # instead of August 1st!
-            now = fields.Datetime.context_timestamp(ir_cron, datetime.utcnow())
-            past_nextcall = fields.Datetime.context_timestamp(
-                ir_cron, fields.Datetime.to_datetime(job['nextcall']))
-
-            # Compute how many call were missed
-            missed_call = past_nextcall
-            missed_call_count = 0
-            while missed_call <= now:
-                missed_call += interval
-                missed_call_count += 1
-            future_nextcall = missed_call
-
-            # Compute how many time we should run the cron
-            effective_call_count = (
-                     1 if not missed_call_count                    # run at least once
-                else 1 if not job['doall']                         # run once for all
-                else missed_call_count if job['numbercall'] == -1  # run them all
-                else min(missed_call_count, job['numbercall'])     # run maximum numbercall times
-            )
-            call_count_left = max(job['numbercall'] - effective_call_count, -1)
+        if completed:
+            ir_cron._mark_as_done(cron_cr, job)
+        elif os.getenv('ODOO_NOTIFY_CRON_CHANGES'): # See: `_notifydb`
+            ir_cron._cr.postcommit.add(ir_cron._notifydb)
 
-            # The actual cron execution
-            for call in range(effective_call_count):
-                ir_cron._callback(job['cron_name'], job['ir_actions_server_id'], job['id'])
+        cron_cr.commit()
 
-        # Update the cron with the information computed above
-        cron_cr.execute("""
+    @classmethod
+    def _execute_job(cls, job, timed_out_counter=0):
```

Comment:
Shouldn't we call this `_run_job` instead ?

Reply:
I'm not a fan of `execute_job` either. What's about `_run_job_in_batch` to hint the presence of a loop?